api/utils/keys/privatekey.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package keys defines common interfaces for Teleport client keys.
+package keys
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"os"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/agent"
+)
+
+const (
+	pkcs8PrivateKeyType = "PRIVATE KEY"
+	pkcs1PrivateKeyType = "RSA PRIVATE KEY"
+	ecPrivateKeyType    = "EC PRIVATE KEY"
+)
+
+// PrivateKey implements crypto.PrivateKey and crypto.Signer, with additional helper methods
+// for performing TLS/SSH handshakes, storing agent keys, and storing private key data.
+type PrivateKey interface {
+	crypto.Signer
+
+	// Equal returns whether the given key is equal to this key
+	Equal(x crypto.PrivateKey) bool
```

Comment:
It seems like we can do `Equal(x PrivateKey) bool` here (with the appropriate changes), it might be worth it for the added type safety.

Reply:
It's not like we're getting real interoperability with `crypto.PrivateKey` anyway, all the `crypto` functions will do type assertions and return an error if it's not one of the three builtin ones.

Reply:
Good point, I think we can just remove `Equal` entirely. We only need to implement crypto.Signer.